enhancements/single-node-developer-cluster-profile.md

```diff
@@ -0,0 +1,643 @@
+  currently used by the SNC tooling to build CRC installation
+  images. Continued evolution of the existing installer, SNC tooling
+  and other CRC components is expected.
+
+* Functional changes for OpenShift. This enhancement proposal does not
+  result in any functional changes to clusters that do not utilize the
+  'single-node-developer' cluster profile.
+
+## Proposal
+
+The 'single-node-developer' cluster profile will be implemented in two
+phases. The first phase, which is targeted for completion for the OCP
+4.7 release, introduces the new cluster profile, and enables some
+simplification of the SNC image build process by utilizing the profile
+to control which components are included in the cluster. The second
+phase, which might begin during the OCP 4.7 development cycle but will
+continue afterwards, leverages the profile to optimize the resource
+utilization of the components included in the cluster, and may further
+simplify the SNC image build process.
+
+### Phase One
+
+In the initial phase, the following annotation will be added to the
+manifest files for all CVO-managed components (operators, deployments,
+etc.) that are to be included in single node developer clusters:
+
+```
+include.release.openshift.io/single-node-developer: "true"
+```
+
+This annotation will need to be added to the manifests for all the
+OpenShift components that are not currently removed by the snc.sh
+script during the CRC build process. The components that are currently
+removed, and thus should not need the annotation added, include:
+
+* most operators running in the openshift-monitoring namespace
+* most operators running in the openshift-insights namespace
+* machine-api-operator
+* machine-config-operator (may need to run once, requiring the annotation)
+* openshift-cloud-credential-operator
+* openshift-cluster-storage-operator
+
+Additional components that are not currently removed by the SNC
+tooling might be also be determined to be unnecessary, and thus not
+require the annotation. Possibilities include:
+
+* openshift-cluster-node-tuning-operator
+* openshift-multus
+
+As an example, the current default manifest for the console downloads
+deployment begins with:
+
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  namespace: openshift-console
+  name: downloads
+  annotations:
+    include.release.openshift.io/self-managed-high-availability: "true"
+spec:
+  replicas: 2
+  selector:
+    # ...
+```
+ 
+After phase one, the same manifest will also apply to the
+'single-node-developer' profile:
+
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  namespace: openshift-console
+  name: downloads
+  annotations:
+    include.release.openshift.io/self-managed-high-availability: "true"
+    include.release.openshift.io/single-node-developer: "true"
+spec:
+  replicas: 2
+  selector:
+    # ...
+```
+
+In order to document that they have been purposely excluded from the
```

Comment:
Is this desirable given that it requires annotating every single manifest for every single operator for every single cluster profile?

Reply:
Hi Maru - Given that the majority of existing manifests are expected to be included in the single-node-developer profile during phase 1, it doesn't seem to have a very high cost, at least for this profile. The benefit is that it enables searching for manifests that don't have any annotation for the profile in order to discover new components that may have been added without considering whether or not they should be included in this profile. Its just a convention intended to simplify maintenance going forward, and not a critical aspect of the enhancement proposal.

Reply:
I suggest dropping this requirement given that searching for the absence of an annotation is as trivial as searching for its negation (e.g. `grep -v [regex]`).

Reply:
We want to be able to differentiate between "this manifest is intentionally excluded from this profile" and "this manifest has no profile annotation because it was forgotten". `grep` won't tell us which it is if we omit the profile annotation in both situations.

Reply:
Why is it desirable to be prescriptive about something that is the really the domain of the component owner - whether manifests should be included or not?

Reply:
The gist of the feedback on this PR so far is that ongoing collaboration will be needed between component teams and the CRC team. These "false" annotations are intended to aid that ongoing collaboration by recording the decisions that have been made (hopefully collaboratively). If there is strong resistance to the idea, we could drop it from the enhancement PR, and leave whether to add these "false" annotations as an implementation detail. But the idea has much more value if it is applied consistently, so that's why I included it in the enhancement.

Reply:
At the moment, there are 1448 YAML manifest files containing one or more resources with the 'self-managed-high-availability' annotation. Of these files, only 350 currently also contain the 'ibm-cloud-managed' annotation. Unless that second number is going to grow a lot closer to the first, I can understand not wanting to add "false" annotations for the 'ibm-cloud-managed' profile to the manifests that are excluded from that profile. So I'm no longer going to suggest that every manifest in every repository should have either a "true" or "false" annotation for every cluster profile, or that this should be checked/enforced by a CI job.

Even if we don't do this consistently across all cluster profiles, adding "false" annotations for the 'single-node-developer' profile to intentionally excluded manifests may still be useful in cases where the same repository also contains manifests included in the profile. In particular, as we start to specialize manifests in phase two, this will make it very clear which version of a manifest applies to which profile (see the examples in this enhancement). But if no manifests in the repository are included in a profile, I'm convinced there is little value adding the "false" annotations for the profile.

Reply:
Changed in upcoming update to only add "false" annotations when other manifests in the same repository are included in the profile.